I've tried this out in Angular InstantSearch, and while it solves most of the issue, ComponentProps didn't work properly yet, experimenting a little further

```
Error: node_modules/@algolia/ui-components-highlight-vdom/dist/esm/Highlight.d.ts:24:53 - error TS2344: Type 'string' does not satisfy the constraint 'never'.

24 export declare type HighlightProps = ComponentProps<'span'> & {
                                                       ~~~~~~
node_modules/@algolia/ui-components-highlight-vdom/dist/esm/Highlight.d.ts:31:155 - error TS2339: Property 'className' does not exist on type '{ classNames?: Partial<HighlightClassNames>; highlightedTagName?: FunctionComponent<any>; nonHighlightedTagName?: FunctionComponent<any>; separator?: ComponentChildren; parts: HighlightedPart[][]; }'.

31 export declare function createHighlightComponent({ createElement, Fragment, }: Renderer): ({ parts, highlightedTagName, nonHighlightedTagName, separator, className, classNames, ...props }: HighlightProps) => JSX.Element;
                                                                                                                                                             ~~~~~~~~~
```

that error shows up because in Angular InstantSearch there's no @types/react, I wonder what the effect would be if it was a regular dependency (do things still override with preact? I don't think so)

--- a/packages/shared/src/Renderer.ts
+++ b/packages/shared/src/Renderer.ts
@@ -2,6 +2,8 @@
 // (e.g., Angular InstantSearch via InstantSearch.js)
 // In the future, this may be fixable by accepting a JSX generic to every type
 // or a `createRenderer` function that implies a JSX generic.
+//
+// This may be removable when this package is only used in an environment with JSX.
 declare global {
   // eslint-disable-next-line @typescript-eslint/no-namespace
   namespace JSX {

--- a/packages/shared/src/Renderer.ts
+++ b/packages/shared/src/Renderer.ts
@@ -12,6 +12,11 @@ declare global {
   }
 }
 
+// For safety when there's no @types/react or preact, we don't directly use JSX.IntrinsicElements
+type IntrinsicElements = keyof JSX.IntrinsicElements extends never
+  ? Record<string, unknown>
+  : JSX.IntrinsicElements;
+
 export type Pragma = (
   type: any,
   props: Record<string, any> | null,
@@ -40,15 +45,15 @@ type FunctionComponent<TProps = {}> = (
 
 export type ElementType<TProps = any> =
   | {
-      [TKey in keyof JSX.IntrinsicElements]: TProps extends JSX.IntrinsicElements[TKey]
+      [TKey in keyof IntrinsicElements]: TProps extends IntrinsicElements[TKey]
         ? TKey
         : never;
-    }[keyof JSX.IntrinsicElements]
+    }[keyof IntrinsicElements]
   | FunctionComponent<TProps>;
 
 export type ComponentProps<
-  TComponent extends keyof JSX.IntrinsicElements
-> = JSX.IntrinsicElements[TComponent];
+  TComponent extends keyof IntrinsicElements
+> = IntrinsicElements[TComponent];
 
 export type VNode<TProps = any> = {
   type: any;

